pkg/metrics/registry.go

```
@@ -0,0 +1,242 @@
+	return r.inner.Register(c)
+}
+
+// Unregister unregisters a collector
+func (r *Registry) Unregister(c prometheus.Collector) bool {
+	return r.inner.Unregister(c)
+}
+
+// goCustomCollectorsRX tracks enabled go runtime metrics.
+var goCustomCollectorsRX = regexp.MustCompile(`^/sched/latencies:seconds`)
+
+// Reinitialize creates a new internal registry and re-registers metrics to it.
+func (r *Registry) Reinitialize() {
+	r.inner = prometheus.NewPedanticRegistry()
+
+	// Default metrics which can't be disabled.
+	r.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{Namespace: Namespace}))
+	r.MustRegister(collectors.NewGoCollector(
+		collectors.WithGoCollectorRuntimeMetrics(
+			collectors.GoRuntimeMetricsRule{Matcher: goCustomCollectorsRX},
+		)))
+	r.MustRegister(newStatusCollector())
+	r.MustRegister(newbpfCollector())
+
+	metrics := make(map[string]metricpkg.WithMetadata)
+	for i, autoMetric := range r.params.AutoMetrics {
+		metrics[autoMetric.Opts().ConfigName] = r.params.AutoMetrics[i]
+	}
+
+	// This is a bodge for a very specific feature, inherited from the old `Daemon.additionalMetrics`.
+	// We should really find a more generic way to handle such cases.
+	metricFlags := r.params.Config.Metrics
+	if r.params.DaemonConfig.DNSProxyConcurrencyLimit > 0 {
+		metricFlags = append(metricFlags, "+"+Namespace+"_"+SubsystemFQDN+"_semaphore_rejected_total")
+	}
+
+	for _, metricFlag := range metricFlags {
+		// This is a temporary hack which allows us to get rid of the centralized metric config without refactoring the
+		// dynamic map pressure registration/unregistion mechanism.
+		// Long term the map pressure metric becomes a smarter component so this is no longer needed.
+		if metricFlag[1:] == "-"+Namespace+"_"+SubsystemBPF+"_map_pressure" {
+			BPFMapPressure = false
+			continue
+		}
+
+		metric := metrics[metricFlag[1:]]
+		if metric == nil {
+			continue
+		}
+
+		switch metricFlag[0] {
```

I think we should add a default case with a warning to this switch, to prevent silently losing some metrics - I _think_ the following would lose the middle metric disable (due to the space before it), as well as silently ignoring the final metric (because there's no plus/minus):

`--metrics "+first_metric, -middle_metric,final_metric"`

I traced `viper.StringSlice` to `encoding/csv` [which doesn't trim whitespace](https://go.dev/play/p/fok-8A8Sdg9).
Given this has been the case in the existing code, probably error/fatal would be breaking and can't be done.